enterprise/coderd/roles.go

```diff
@@ -12,70 +14,86 @@ import (
 	"github.com/coder/coder/v2/codersdk"
 )
 
-// patchRole will allow creating a custom role
-//
-// @Summary Upsert a custom site-wide role
-// @ID upsert-a-custom-site-wide-role
-// @Security CoderSessionToken
-// @Produce json
-// @Tags Members
-// @Success 200 {array} codersdk.Role
-// @Router /users/roles [patch]
-func (api *API) patchRole(rw http.ResponseWriter, r *http.Request) {
-	ctx := r.Context()
+type enterpriseCustomRoleHandler struct {
+	Enabled bool
+}
 
-	var req codersdk.Role
-	if !httpapi.Read(ctx, rw, r, &req) {
-		return
+func (h enterpriseCustomRoleHandler) PatchOrganizationRole(ctx context.Context, db database.Store, rw http.ResponseWriter, orgID uuid.UUID, role codersdk.Role) (codersdk.Role, bool) {
+	if !h.Enabled {
+		httpapi.Write(ctx, rw, http.StatusForbidden, codersdk.Response{
+			Message: "Custom roles is not enabled",
+		})
+		return codersdk.Role{}, false
 	}
 
-	if err := httpapi.NameValid(req.Name); err != nil {
+	// Only organization permissions are allowed to be granted
+	if len(role.SitePermissions) > 0 {
 		httpapi.Write(ctx, rw, http.StatusBadRequest, codersdk.Response{
-			Message: "Invalid role name",
-			Detail:  err.Error(),
+			Message: "Invalid request, not allowed to assign site wide permissions for an organization role.",
+			Detail:  "organization scoped roles may not contain site wide permissions",
 		})
-		return
+		return codersdk.Role{}, false
 	}
 
-	if len(req.OrganizationPermissions) > 0 {
-		// Org perms should be assigned only in org specific roles. Otherwise,
-		// it gets complicated to keep track of who can do what.
+	if len(role.UserPermissions) > 0 {
 		httpapi.Write(ctx, rw, http.StatusBadRequest, codersdk.Response{
-			Message: "Invalid request, not allowed to assign organization permissions for a site wide role.",
-			Detail:  "site wide roles may not contain organization specific permissions",
+			Message: "Invalid request, not allowed to assign user permissions for an organization role.",
+			Detail:  "organization scoped roles may not contain user permissions",
 		})
-		return
+		return codersdk.Role{}, false
+	}
+
+	if len(role.OrganizationPermissions) > 1 {
+		httpapi.Write(ctx, rw, http.StatusBadRequest, codersdk.Response{
+			Message: "Invalid request, Only 1 organization can be assigned permissions",
+			Detail:  "roles can only contain 1 organization",
+		})
+		return codersdk.Role{}, false
+	}
+
+	if len(role.OrganizationPermissions) == 1 {
+		_, exists := role.OrganizationPermissions[orgID.String()]
+		if !exists {
+			httpapi.Write(ctx, rw, http.StatusBadRequest, codersdk.Response{
+				Message: fmt.Sprintf("Invalid request, expected permissions for only the organization %q", orgID.String()),
+				Detail:  fmt.Sprintf("only org id %s allowed", orgID.String()),
+			})
+			return codersdk.Role{}, false
+		}
```

Comment:
This check becomes redundant if `OrganizationPermissions` is just a `[]Permission` instead of a `map[OrgID][]Permission`